call.go

```diff
@@ -202,92 +201,95 @@ func invoke(ctx context.Context, method string, args, reply interface{}, cc *Cli
 		Last:  true,
 		Delay: false,
 	}
+	callHdr := &transport.CallHdr{
+		Host:   cc.authority,
+		Method: method,
+	}
+	if cc.dopts.cp != nil {
+		callHdr.SendCompress = cc.dopts.cp.Type()
+	}
+	if c.creds != nil {
+		callHdr.Creds = c.creds
+	}
+	firstAttempt := true
+
 	for {
-		var (
-			err    error
-			t      transport.ClientTransport
-			stream *transport.Stream
-			// Record the done handler from Balancer.Get(...). It is called once the
-			// RPC has completed or failed.
-			done func(balancer.DoneInfo)
-		)
-		// TODO(zhaoq): Need a formal spec of fail-fast.
-		callHdr := &transport.CallHdr{
-			Host:   cc.authority,
-			Method: method,
-		}
-		if cc.dopts.cp != nil {
-			callHdr.SendCompress = cc.dopts.cp.Type()
-		}
-		if c.creds != nil {
-			callHdr.Creds = c.creds
+		// Check to make sure the context has expired.  This will prevent us from
+		// looping forever if an error occurs for wait-for-ready RPCs where no data
+		// is sent on the wire.
+		select {
```

Comment:
there's no reason to use a select here, and doing so can result in worse performance than the simpler alternative:
```
if err := ctx.Err(); err != nil {
  return toRPCErr(err)
}
```

Reply:
It turns out this is not true.  I made some benchmarks, and found the following results:

```
BenchmarkCancelContextErrNoErr-12            	100000000	        52.2 ns/op
BenchmarkCancelContextErrGotErr-12           	100000000	        51.9 ns/op
BenchmarkCancelContextChannelNoErr-12        	200000000	        19.9 ns/op
BenchmarkCancelContextChannelGotErr-12       	100000000	        36.1 ns/op
```

Using the channel is faster as of today.  I believe the problem is the `defer` in `context.Err()`:

https://github.com/golang/net/blob/cd69bc3fc700721b709c3a59e16e24c67b58f6ff/context/pre_go17.go#L162

These numbers line up well with our benchmarks for mutexes vs. channels and defer vs no-defer:
```
BenchmarkMutexWithDefer-12                	20000000	        51.2 ns/op
BenchmarkMutex-12                         	100000000	        14.5 ns/op
BenchmarkSelectClosed-12                  	50000000	        20.3 ns/op
BenchmarkSelectOpen-12                    	300000000	         5.74 ns/op
```

Note how the select is faster than the mutex with a defer, but slower than the mutex with an inline `Unlock()`.

Reply:
Quite surprising, thanks for the details. Your comment points to pre_go17.go but I assume these benchmarks were run on 1.9?

Also interesting that selecting on the open channel is faster than the mutex with inline `Unlock`.

Reply:
By the way, there's something fishy here - how can `BenchmarkCancelContextChannelGotErr` ever be faster than `BenchmarkCancelContextErrGotErr`? The former does strictly more work.

Reply:
> Quite surprising, thanks for the details. Your comment points to pre_go17.go but I
> assume these benchmarks were run on 1.9?

Yes -- presumably the std context package is implemented identically (based on the results that makes sense).

> The former does strictly more work.

Good catch.  Actually, it does not do strictly more work, because I forgot to call Err() in this case.  Adding it results in:

```
BenchmarkCancelContextErrGotErr-12        	100000000	        52.4 ns/op
BenchmarkCancelContextChannelGotErr-12    	50000000	        94.8 ns/op
BenchmarkTimerContextErrGotErr-12         	100000000	        52.5 ns/op
BenchmarkTimerContextChannelGotErr-12     	50000000	        86.7 ns/op
```

I would still lean toward the existing approach, as it optimizes the non-error case.  But maybe the Go team should think about removing that defer...

Reply:
cc @Sajmani -- looks like that defer has been there since inception; perhaps it's worth replace with the an inline `Unlock` as @dfawley has shown.

Reply:
FWIW I filed https://github.com/golang/go/issues/22401 for this.